src/classes/RD2_RecurringDonations_TDTM.cls

```diff
@@ -260,38 +260,101 @@ public class RD2_RecurringDonations_TDTM extends TDTM_Runnable {
+
+            for (SObject sobj : revisedSchedules.objectsToUpdate) {
+                addUpdatedSchedules(sobj, newSchedulesByRdId);
+            }
+
+            for (SObject sobj : revisedSchedules.objectsToInsert) {
+                addUpdatedSchedules(sobj, newSchedulesByRdId);
+            }
+        }
+
+        // Merge the schedules from RDs NOT needing schedule updates with the updated schedules
+        for (Id rdId : newSchedulesByRdId.keySet()) {
+            schedulesByRdId.put(rdId, newSchedulesByRdId.get(rdId));
+        }
 
         for (npe03__Recurring_Donation__c rd : rds) {
-            rd.npe03__Next_Payment_Date__c = scheduleService.getNextDonationDate(rd);
+            rd.npe03__Next_Payment_Date__c = scheduleService.getNextDonationDate(schedulesByRdId.get(rd.Id));
         }
     }
 
     /***
-    * @description Handles schedule modification/creation after modification of recurring donation
-    * @param rds Trigger.new records
-    * @param oldRds Trigger.old records
+    * @description Retrieves recurring donation schedules associated with recurring donation list
+    * @param rds List<npe03__Recurring_Donation__c>
     */
-    private void synchronizeSchedulesAfterRDUpdate(List<npe03__Recurring_Donation__c> rds, List<npe03__Recurring_Donation__c> oldRds) {
-
-        Map<Id, npe03__Recurring_Donation__c> filteredRecurringDonations = getRecurringDonationsNeedingScheduleUpdates(rds, oldRds);
+    private Map<Id, List<RecurringDonationSchedule__c>> getExistingSchedules(List<npe03__Recurring_Donation__c> rds) {
 
         Map<Id, List<RecurringDonationSchedule__c>> oldSchedules = new Map<Id, List<RecurringDonationSchedule__c>>();
+
         for (RecurringDonationSchedule__c oldSchedule : [
-                SELECT RecurringDonation__c, EndDate__c FROM RecurringDonationSchedule__c
-                WHERE RecurringDonation__c IN :filteredRecurringDonations.keySet()
+                SELECT RecurringDonation__c, StartDate__c, EndDate__c, InstallmentFrequency__c, InstallmentPeriod__c, DayOfMonth__c FROM RecurringDonationSchedule__c
+                WHERE RecurringDonation__c IN :rds
                 AND Active__c = true]) {
             if (!oldSchedules.containsKey(oldSchedule.RecurringDonation__c)) {
                 oldSchedules.put(oldSchedule.RecurringDonation__c, new List<RecurringDonationSchedule__c>());
             }
             oldSchedules.get(oldSchedule.RecurringDonation__c).add(oldSchedule);
         }
 
+        return oldSchedules;
+    }
+
+    /***
+    * @description Adds recurring donation schedules to map by recurring donation Id
+    * @param sobj SObject
+    * @param newSchedulesByRdId Map<Id, List<RecurringDonationSchedule__c>>
+    */
+    private void addUpdatedSchedules(SObject sobj, Map<Id, List<RecurringDonationSchedule__c>> newSchedulesByRdId) {
+
+        if (sobj.getSObjectType() == RecurringDonationSchedule__c.SObjectType) {
```

Comment:
Since this is a private method in this class, can this ever be any other type than a `RecurringDonationSchedule__c` record, especially if the first parameter is of `RecurringDonationSchedule__c` type instead of SObject?

I think you can do that by changing the calling methods to 
`addUpdatedSchedules((RecurringDonationSchedule__c)sobj, newSchedulesByRdId); `

Reply:
Right now it is true that the only type of sobject this can be is a `RecurringDonationSchedule__c` record.  I was a bit reluctant to lock that conclusion into code, but that degree of caution may be overkill.